lib/enigma.rb

```diff
@@ -83,4 +107,48 @@ def create_shifts(key, date)
     split_key(key).zip(date_to_offsets(date)).map { |nums| nums.reduce(:+) }
   end
 
+  def find_key(encrypted_message, date)
+    key  = ""
+    offsets = date_to_offsets(date)
+    find_shifts(encrypted_message, date).each_with_index do |shift, index|
+      if ((shift - offsets[index]) % 27).to_s.length == 1 && index == 0
+        key.concat("0" + ((shift - offsets[index]) % 27).to_s)
+      elsif index == 0
+        key.concat(((shift - offsets[index]) % 27).to_s)
+      else
+        new_key_value = 0
+        until key.length > index + 1
+          if (key.chars.last.concat(new_key_value.to_s)).to_i % 27 == shift
+            key.concat((((new_key_value - offsets[index]) % 27).to_s))
+          end
+          new_key_value += 1
+        end
+      end
+    end
+    key
+  end
+
+  def find_shifts(encrypted_message, date)
+    shifts = []
+    split_encrypted = split_message(encrypted_message)
+    last_chars_length = split_encrypted.last.length
+    if last_chars_length == 4
+      shifts = split_encrypted.last.zip([" ", "e", "n", "d"]).map do |encrypted_char, end_char|
+        shift_amount(encrypted_char, end_char)
+      end
+    elsif last_chars_length < 4
+      shifts = split_encrypted.last.zip([" ", "e", "n", "d"].last(last_chars_length)).map do |encrypted_char, end_char|
+        shift_amount(encrypted_char, end_char)
+      end
+      shifts << split_encrypted[-2].last(4 - last_chars_length).each_with_index.map do |encrypted_char, index|
+        shift_amount(encrypted_char, [" ", "e", "n", "d"][index])
+      end
+    end
+    shifts.flatten
```

Comment:
This method needs to be refactored as well. If the message length is a multiple of 4 it isn't hard to find the shifts because you can just compare the last 4 characters to the last 4 encrypted. If it isn't a multiple of 4 you have to go into the second to last set of characters in the encrypted message to find the rest of the shifts.